Hw3

BaranovaElena opened the pull request:
чую напортачила с mvvm, не уверена что это правильно - создавать appState для каждого экрана, но по-другому не знаю как подгрузить только нужные мне данные, а не всё подряд

![fbD01hWFBAQ](https://user-images.githubusercontent.com/75544888/126777211-add25eab-0f75-4b55-9973-8815122cbf2b.jpg)
![XwRLJL0xYag](https://user-images.githubusercontent.com/75544888/126777215-672ee6f5-509f-4089-9e40-b3598403bae0.jpg)
![JVC7Ut1Cmh0](https://user-images.githubusercontent.com/75544888/126777217-5f9a59a2-a63a-4078-a959-208c8f441596.jpg)
![MS0Fjmv1kYc](https://user-images.githubusercontent.com/75544888/126777219-74b0b8c7-b750-4cfd-bf79-c3be69f03efe.jpg)

BaranovaElena committed:
в модель добавлена функция выдачи двух списков погод

BaranovaElena committed:
в модель добавлена функция выдачи двух списков погод

BaranovaElena committed:
добавлен recycler с переключением двух списков по кнопке

BaranovaElena committed:
передача данных через bundle от адаптера в фрагмент с погодой, обработка клика по городу

BaranovaElena committed:
передача данных через bundle от адаптера в фрагмент с погодой, обработка клика по городу

BaranovaElena committed:
ui переделан под 2 списка, создан второй адаптер но без обработки

BaranovaElena committed:
ui переделан под 2 списка, создан второй адаптер но без обработки

BaranovaElena committed:
сделаны два списка на экране

BaranovaElena committed:
сделаны два списка на экране

BaranovaElena committed:
добавлена навигация. стартовым экраном сделала детальное отображение погоды в москве
DetailsFragment берет погоду в москве по умолчанию не из viewModel, надо исправить

BaranovaElena committed:
добавлена навигация. стартовым экраном сделала детальное отображение погоды в москве
DetailsFragment берет погоду в москве по умолчанию не из viewModel, надо исправить

BaranovaElena committed:
добавлены свои viewModel и appState для DetailsFragment
DetailsFragment берет погоду в Москве по умолчанию из viewModel

BaranovaElena committed:
добавлены свои viewModel и appState для DetailsFragment
DetailsFragment берет погоду в Москве по умолчанию из viewModel

BaranovaElena committed:
локализация

BaranovaElena committed:
темная тема

BaranovaElena committed:
рефакторинг кода:
применение let apply,
lazy для viewModel,
замена null-проверок на ?. ?: ,
сокращение лямбд,
функция-расширение для snackBar

BaranovaElena committed:
добавила сердечко на катрочку с городом, потом планирую делать список любимых городов в окне favorites

BaranovaElena committed:
добавила больше информации на экран с детальным отображением погоды

BaranovaElena committed:
hw5: сделана загрузка погоды с api (некрасиво прямо во фрагменте, посмотреть что вообще заработает)

BaranovaElena committed:
hw5: загрузка вынесена в отдельный класс, добавлены извлекаемые данные

BaranovaElena committed:
hw5: обработка ошибки загрузки

BaranovaElena committed:
hw3: исправления после rebase ветки hw2 с новым коммитом

BaranovaElena committed:
hw4: исправления после rebase ветки hw3 с новым коммитом

BaranovaElena committed:
hw5: привожу в порядок модель и репозитории:
DetailsFragment переделала на mvvm
viewModel не зависит от реализации репозитория WeathersRepo
загрузка из инета переместилась в WeathersRepoImplApi
weather содержит WeatherDTO, чтобы не было разногласий между загрузкой с api и заглушкой

BaranovaElena committed:
hw7-refactor-packages: перераспределила все по папкам по примеру как в чате обсуждали

BaranovaElena committed:
hw7: retrofit

BaranovaElena committed:
hw7: загрузка svg иконок погоды через GlideToVectorYou

BaranovaElena committed:
hw7: загрузка картинок городов в список через Glide

BaranovaElena committed:
hw8: добавлен экран с историей запросов, история хранится в базе Room

BaranovaElena committed:
hw9: добавлен запрос на gps, при запуске выводится погода в текущем местоположении

kshalnov commented on app/src/main/java/com/example/weather/model/Repo.kt:
```diff
@@ -2,5 +2,6 @@ package com.example.weather.model
 interface Repo {
     fun getWeatherFromServer(): Weather
-    fun getWeatherFromLocalStorage(): Weather
+    fun getWeatherFromLocalStorageRus(): List<Weather>
```
допустим теперь надо сделать так чтобы приложение работало в США, какие изменения придётся внести?

BaranovaElena replied:
во-первых использовать для отображения городов репозиторий, возвращающий не погоду, а только города. так мы абстрагируемся от загрузки кучи данных их интернета при отображении списка городов.
во-вторых сделать интерфейс с функциями getCitiesListWorld() и getCitiesListLocal() и разные реализации функции getCitiesListLocal() под разные локализации. а какую использовать пусть viewmodel решает

kshalnov commented on app/src/main/java/com/example/weather/ui/viewmodel/CityListViewModel.kt:
```diff
@@ -6,19 +6,24 @@ import androidx.lifecycle.ViewModel
-class MainViewModel : ViewModel() {
-    private val liveLoadStateValue: MutableLiveData<LoadState> = MutableLiveData()
+class CityListViewModel : ViewModel() {
+    private val liveLoadStateValue: MutableLiveData<LoadAllCitiesState> = MutableLiveData()
```
loadStateLiveData я бы как-то так назвал

BaranovaElena replied:
ок

kshalnov commented on app/src/main/java/com/example/weather/view/CityListAdapter.kt:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CityListViewHolder {
+        return CityListViewHolder(
+            LayoutInflater.from(parent.context)
+                .inflate(R.layout.city_list_item, parent, false) as View
+        )
```
в прошлом курсе инфлейтили внутри холдера, а почему теперь не так?

BaranovaElena replied:
я уже потом поняла, что не все что в методичке значит правильно)

kshalnov commented on app/src/main/java/com/example/weather/view/CityListAdapter.kt:
```diff
@@ -0,0 +1,51 @@
+    inner class CityListViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+
+        fun bind(weather: Weather) {
+            itemView.findViewById<TextView>(R.id.city_list_item_text_view).text = weather.city.name
```
может биндинг приделать?

BaranovaElena replied:
мне показалось что это нецелесообразно ради одного текстового поля делать

kshalnov commented on app/src/main/java/com/example/weather/view/CityListFragment.kt:
```diff
@@ -0,0 +1,101 @@
+import com.google.android.material.snackbar.Snackbar
+
+class CityListFragment : Fragment() {
+    private var binding: FragmentCityListBinding? = null
```
стоит сделать "backing field" или инициализировать через делегат, чтобы потом вечно !! или ? не ставить

BaranovaElena replied:
спасибо

kshalnov commented on app/src/main/java/com/example/weather/view/DetailsFragment.kt:
```diff
@@ -0,0 +1,94 @@
+    companion object {
+        const val BUNDLE_EXTRA_KEY = "WEATHER_BUNDLE_EXTRA_KEY"
+
+        fun newInstance(bundle: Bundle): DetailsFragment {
```
теперь если мне нужно показать этот фрагмент мне надо зайти в код, посмотреть ключи которые используются, найти в коде какие объекты нужны. Нарушена инкапсуляция. Надо аргументами метода передавать, а не бандл

BaranovaElena replied:
только вчера смотрела лекцию, где вы эту ошибку упоминали. исправлю

BaranovaElena committed:
hw10: hw3: исправила замечания к дз3
сделан интерфейс CityListsRepo, реализации которого содержат списки городов в разных локализациях. CityListFragment.kt теперь использует его вместо репозитоирия, содержащего такие списки с погодой
исправлена ошибка с использованием bundle для детального фрагмента внутри фрагмента со списком. сделан метод newInstance(city) и от фрагмента к фрагменту через mainActivity передается city

BaranovaElena committed:
hw10: сделала проверку для gps если он выключен

BaranovaElena committed:
hw10: подключена карта, переставляется маркер, по маркеру выводится название города, по кнопке "показать погоду здесь" открывается окно с погодой в отмеченной маркером точке

BaranovaElena committed:
hw10: добавлен ввод адреса, отображение координат введенного адреса

BaranovaElena committed:
hw10: навела порядок в strings.xml и dimens.xml

BaranovaElena committed:
hw11: подключила уведомления из firebase, сделала разные иконки для трех типов уведомлений

BaranovaElena committed:
hw11: сделала уведомление по нажатию иконки в action bar

BaranovaElena committed:
Merge pull request #11 from BaranovaElena/hw11-notify_btn

Hw11 - push notifications

BaranovaElena committed:
Merge pull request #10 from BaranovaElena/hw10

Hw10 - maps and geolocation

BaranovaElena committed:
Merge pull request #13 from BaranovaElena/hw9

Hw9 - content provider and location permission

BaranovaElena committed:
Merge pull request #8 from BaranovaElena/hw8

Hw8 - Room

BaranovaElena committed:
Merge pull request #7 from BaranovaElena/hw7

Hw7 - Glide

BaranovaElena committed:
Merge pull request #14 from BaranovaElena/hw7-refactor-packages

Hw7 refactor packages

BaranovaElena committed:
Merge pull request #5 from BaranovaElena/hw5

Hw5 - http

BaranovaElena committed:
Merge pull request #4 from BaranovaElena/hw4

Hw4
